R/tax_unique.R

```diff
@@ -0,0 +1,279 @@
+#' Filter occurrences to unique taxa
+#'
+#' A function to filter a list of taxonomic occurrences to unique taxa of a
+#' predefined resolution. In many cases palaeobiologists achieve this by
+#' retaining only unique occurrences identified to the given taxonomic
+#' resolution, however here we also retain occurrences identified to a coarser
+#' resolution which are not already represented within the dataset.
+#' If we take the following set of occurrences:
+#' Albertosaurus sarcophagus
+#' Ankylosaurus sp.
+#' Aves indet.
+#' Ceratopsidae indet.
+#' Hadrosauridae indet.
+#' Ornithominus sp.
+#' Tyrannosaurus rex
+#' then a filter for species-level identifications will reduce the species
+#' diversity to two. However, none of these clades are nested within another, so
+#' each of the indeterminately identified occurrences represents at least one
+#' species not already represented in the dataset. This function is designed to
+#' deal with such taxonomic data, and would retain all seven 'species' in this
+#' dataset.
+#' Data can be supplied in one of two forms, either a dataframe containing
+#' standard occurrence data from the Paleobiology Database, or a series of
+#' vectors which each describe a different taxonomic level for a set of
+#' occurrences (the order of names in the vectors must correspond). At a
+#' minimum, species, genus, and family vectors must be supplied. Occurrence
+#' data can be filtered to retain either unique species or unique genera, and
+#' can be defined as unique either across the whole dataset or within set
+#' categories (e.g. temporal intervals, spatial bins). Missing data should be
+#' indicated with NAs.
+#'
+#' @param paleobioDB \code{dataframe}. A dataframe of taxonomic occurrences
+#' downloaded directly from the Paleobiology Database. The dataframe must
+#' include the following columns: class, order, family, genus, accepted_name.
+#' @param species \code{character}. A vector of species names.
+#' @param genus \code{character}. A vector of genus names.
+#' @param family \code{character}. A vector of family names.
+#' @param order \code{character}. A vector of order names.
+#' @param class \code{character}. A vector of class names.
+#' @param resolution \code{character}. The taxonomic resolution at which to
+#' identify unique occurrences, either species (the default) or genera.
+#' @param by \code{numeric, character}. A category within which to determine
```

Comment:
Would it be possible to have multiple grouping categories? e.g., country and time interval?

Reply:
Alternatively, you could provide an example showing how a user can concat two variables into one column? This would then work the same?

Reply:
We've had this conversation about other functions as it's a common problem - I would be in favour of writing our functions around a single grouping variable and providing a helper function (or guidance on how to use external functions) to convert multiple grouping variables into a single one.

Reply:
That could work, but then users would need to unconcatenate those variables after, right?

Reply:
In base R you can split a data.frame then use lapply to apply a function to each part, e.g.:
```r
lst <- split(tetrapods, f = ~ cc + formation)
results <- do.call(rbind, lapply(lst, function(item) { blah }))
```

Reply:
Shower thought: another option is to remove this functionality entirely and put it on the user to apply this function for their desired groups. This enables them to use base R (like above) or tidyverse (group_by + map_group, or whatever) depending on their comfort. If you go this route, we should probably include an example that shows how this could be done (in base R).

Reply:
I see the appeal / ease of leaving users to split their datasets, but in some cases this could be really labour-intensive, and feels dissatisfactory given that it's such a common problem. But I also agree that back-converting the grouping codes could be hard work, especially if there isn't a 1:1 row match with the original dataset. Using e.g. leftjoin() would be fairly easy for illustrating an example (the codes can be added as a row to the old dataset then used as labels in the new), but I'm not sure how difficult it would be without using tidyverse functions?

Reply:
You can split the variables easily in base:

```
conc <- c("bin1_collection1", "bin1_collection2")
conc <- strsplit(x = conc, split = "_")
do.call(rbind.data.frame, conc)
```

I guess if we returned the original occdf with an appended unique species column, you wouldn't even need to do that?